_posts/2019-07-10-feature-columns/feature_columns.Rmd

```diff
@@ -0,0 +1,566 @@
+  layer_dropout(rate = 0.25) %>%
+  layer_dense(units = 10, activation = "relu", kernel_regularizer = regularizer_l2(l)) %>%
+  layer_dropout(rate = 0.25) %>%
+  layer_dense(units = 5, activation = "relu", kernel_regularizer = regularizer_l2(l)) %>%
+  layer_dropout(rate = 0.25) %>%
+  layer_dense(units = 1, activation = "sigmoid", kernel_regularizer = regularizer_l2(l))
+
+model <- keras_model(inputs, output)
+```
+
+How do we feed this model? In the non-feature-columns example, we would have had to feed each input separately, passing a list of tensors. Now we can just pass it the complete training set all at once:
+
+```{r}
+model %>% fit(x = training, y = training$target)
+```
+
+In the Kaggle competition, submissions are [evaluated using the normalized Gini coefficient](https://www.kaggle.com/c/porto-seguro-safe-driver-prediction/overview/evaluation), which we can calculate with the help of a new metric available in Keras, `tf$keras$metrics$AUC()`. For training, we can use an approximation to the AUC due to Yan et al. (2003) [@yan2003optimizing]. Then training is as straightforward as:
+
+```{r}
+auc <- tf$keras$metrics$AUC()
+
+gini <- custom_metric(name = "gini", function(y_true, y_pred) {
+  2*auc(y_true, y_pred) - 1
+})
+
+# Yan, L., Dodier, R., Mozer, M. C., & Wolniewicz, R. (2003). 
+# Optimizing Classifier Performance via an Approximation to the Wilcoxon-Mann-Whitney Statistic.
+roc_auc_score <- function(y_true, y_pred) {
+
+  pos = tf$boolean_mask(y_pred, tf$cast(y_true, tf$bool))
+  neg = tf$boolean_mask(y_pred, !tf$cast(y_true, tf$bool))
+
+  pos = tf$expand_dims(pos, 0L)
+  neg = tf$expand_dims(neg, 1L)
+
+  # original paper suggests performance is robust to exact parameter choice
+  gamma = 0.2
+  p     = 3
+
+  difference = tf$zeros_like(pos * neg) + pos - neg - gamma
+
+  masked = tf$boolean_mask(difference, difference < 0.0)
+
+  tf$reduce_sum(tf$pow(-masked, p))
+}
+
+model %>%
+  compile(
+    loss = roc_auc_score,
+    optimizer = optimizer_adam(),
+    metrics = list(auc, gini)
+  )
+
+model %>%
+  fit(
+    x = training,
+    y = training$target,
+    epochs = 50,
+    validation_data = list(testing, testing$target),
+    batch_size = 512
+  )
+
+predictions <- predict(model, testing)
+Metrics::auc(testing$target, predictions)
+```
+
+After 50 epochs, we achieve an AUC of 0.64 on the validation set, or equivalently, a Gini coefficient of 0.27. Not a bad result for a simple fully connected network!
+
+We've seen how using feature columns automates away a number of steps in setting up the network, so we can spend more time on actually tuning it. This is most impressively demonstrated on a dataset like this, with more than a handful categorical columns. However, to explain a bit more what to pay attention to when using feature columns, it's better to choose a smaller example where we can easily do some peeking around.
+
+Let's move on the second application. 
```

Comment:
```suggestion
```

Reply:
hm, not seeing the "green" alternative here, adding the "the" in manually...